feat(opentelemetry-sdk-node): automatically configure metrics exporter based on environment variables

bhaskarbanerjee opened the pull request:
<!--
We appreciate your contribution to the OpenTelemetry project! 👋🎉

Before creating a pull request, please make sure:
- Your PR is solving one problem
- Please provide enough information so that others can review your pull request
- You have read the guide for contributing
  - See https://github.com/open-telemetry/opentelemetry-js/blob/main/CONTRIBUTING.md
- You signed all your commits (otherwise we won't be able to merge the PR)
  - See https://github.com/open-telemetry/community/blob/master/CONTRIBUTING.md#sign-the-cla
- You added unit tests for the new functionality
- You mention in the PR description which issue it is addressing, e.g. "Fixes #xxx". This will auto-close
  the issue that your PR fixes (if such)
-->

## Which problem is this PR solving?

Provides ability to auto instrument metrics based on environment variables

Fixes #4551

## Short description of the changes

## Type of change

Please delete options that are not relevant.

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [ ] Test A

## Checklist:

- [x] Followed the style guidelines of this project
- [x] Unit tests have been added
- [x] Documentation has been updated

dyladan commented on experimental/packages/opentelemetry-sdk-node/src/sdk.ts:
```diff
@@ -334,6 +350,99 @@ export class NodeSDK {
     );
   }
 
+  public configureMetricProviderFromEnv(): MetricReader[] {
```
This doesn't need to be a method on the class because it isn't accessing `this`. Making it a function improves minification and tree-shaking. It looks like the function wouldn't need to be exported too, keeping the external API surface as small as possible. It should at least not be public.

bhaskarbanerjee replied:
done

bhaskarbanerjee replied:
Done

dyladan commented on experimental/packages/opentelemetry-sdk-node/src/sdk.ts:
```diff
@@ -400,4 +509,11 @@ export class NodeSDK {
+  /**
+   * @Returns param value, if set else returns the default value
+   */
+  private getValueInMillis(envName: string, defaultValue: number): number {
```
This could just be a function because it doesn't access `this`

dyladan commented on experimental/packages/opentelemetry-sdk-node/tsconfig.json:
```diff
@@ -39,9 +39,6 @@
     {
       "path": "../../../semantic-conventions"
     },
-    {
-      "path": "../api-logs"
-    },
```
Did you mean to remove this?

bhaskarbanerjee replied:
No, that was a mistake

bhaskarbanerjee replied:
reverted

JacksonWeber commented on experimental/packages/opentelemetry-sdk-node/README.md:
```diff
@@ -212,6 +212,9 @@ This is an alternative to programmatically configuring an exporter or span proce
 | OTEL_EXPORTER_OTLP_TRACES_PROTOCOL  | The transport protocol to use on OTLP trace requests. Options include `grpc`, `http/protobuf`, and `http/json`. Default is `http/protobuf`.                  |
 | OTEL_EXPORTER_OTLP_METRICS_PROTOCOL | The transport protocol to use on OTLP metric requests. Options include `grpc`, `http/protobuf`, and `http/json`. Default is `http/protobuf`.                 |
 | OTEL_EXPORTER_OTLP_LOGS_PROTOCOL    | The transport protocol to use on OTLP log requests. Options include `grpc`, `http/protobuf`, and `http/json`. Default is `http/protobuf`.                    |
+| OTEL_METRICS_EXPORTER    | Metrics exporter to be used. options are `otlp`, `prometheus`, `console` or `none`   
```
```suggestion
| OTEL_METRICS_EXPORTER    | Metrics exporter to be used. Options are `otlp`, `prometheus`, `console` or `none`   
```

JacksonWeber commented on experimental/packages/opentelemetry-sdk-node/src/sdk.ts:
```diff
@@ -334,6 +350,99 @@ export class NodeSDK {
+    // ONLY if OTEL_METRICS_EXPORTER is set
+    if (enabledExporters.length === 0) {
+      diag.info('OTEL_METRICS_EXPORTER is empty. Using default otlp exporter.');
+      // enabledExporters.push('otlp');
```
Clean up stray commented line of code.

bhaskarbanerjee replied:
Done

pichlermarc commented on experimental/packages/opentelemetry-sdk-node/src/sdk.ts:
```diff
@@ -334,6 +350,99 @@ export class NodeSDK {
+      metricsExporterList.split(',')
+    );
+
+    // ONLY if OTEL_METRICS_EXPORTER is set
+    if (enabledExporters.length === 0) {
```
Hmm, this will always log though even if `OTEL_METRICS_EXPORTER` is undefined. I think it'd probably the best to check if `process.env.OTEL_METRIC_EXPORTER` is `undefined` and returning an empty array right at the beginning of the function to avoid that any log spurious log messages pop up that are not applicable to what the user actually configured.

bhaskarbanerjee replied:
Done

pichlermarc commented on experimental/packages/opentelemetry-sdk-node/test/sdk.test.ts:
```diff
@@ -1095,6 +1101,218 @@ describe('Node SDK', () => {
+        sharedState.metricCollectors[0]._metricReader._exporter instanceof
+          OTLPGrpcMetricExporter
+      );
+      assert(
+        sharedState.metricCollectors[0]._metricReader._exportInterval === 60000
+      );
```
When checking for strict equality, you can use this
```suggestion
      assert.strictEqual(sharedState.metricCollectors[0]._metricReader._exportInterval, 60000);
```

this'll give more useful error messages if the test fails, please also align the other tests you added where possible.

pichlermarc reviewed (commented):
thanks for working on this 🙌

pichlermarc commented:
Please also [sign the CLA](https://github.com/open-telemetry/opentelemetry-js/pull/5168#issuecomment-2479962309), otherwise we'll not be able to proceed with this PR. 🙂

bhaskarbanerjee commented:
@dyladan @pichlermarc @JacksonWeber Thanks for you review, I have updated the PR, seeking feedback.

pichlermarc reviewed (commented):
Thanks for addressing the comments 🙌 

Thanks for signing the CLA. 👍  Looks like there's still one commit that does not have the CLA signed (is it using different e-mail, perhaps 🤔) - having all commits authorized by the CLA is a required check, so we'll not be able to merge this PR without that. Would you mind having another look? 🙂

bhaskarbanerjee commented:
@pichlermarc 
I am unsure how the commit https://github.com/open-telemetry/opentelemetry-js/commit/73c20718d6e9d1260ba302b63bbb28b08d5a52bf made it from user bhskrbnrj

Those changes are mine but that is not my user id. Let me see if I can remove that commit from my PR.

bhaskarbanerjee commented:
That PR is not from my fork (might be I was fiddling with the PR from UI). The use name seems automated (i.e. stripped off all vowels from my name). Are you OK if I raise a new PR (by including all of these PR comments in there? 
I will keep this PR open until then. and once the other PR is merged, I will close this out.

Please advice @pichlermarc

bhaskarbanerjee commented on experimental/packages/opentelemetry-sdk-node/src/sdk.ts:
```diff
@@ -86,6 +96,113 @@ export type LoggerProviderConfig = {
+    diag.info('OTEL_METRICS_EXPORTER is empty. Using default otlp exporter.');
+  }
+
+  if (enabledExporters.includes('none')) {
+    diag.info(
+      `OTEL_METRICS_EXPORTER contains "none". Metric provider will not be initialized.`
+    );
```
sure

bhaskarbanerjee commented:
I had to rollback all commits in this PR because the first commit was bad (not from a legit fork). I shall soon update this PR with the latest code. @pichlermarc

bhaskarbanerjee commented:
@pichlermarc I have cleaned up the bad PR and addressed all review comments. Seeking your review on this commit

pichlermarc commented on experimental/packages/opentelemetry-sdk-node/README.md:
```diff
@@ -212,6 +212,9 @@ This is an alternative to programmatically configuring an exporter or span proce
 | OTEL_EXPORTER_OTLP_METRICS_PROTOCOL | The transport protocol to use on OTLP metric requests. Options include `grpc`, `http/protobuf`, and `http/json`. Default is `http/protobuf`.                 |
 | OTEL_EXPORTER_OTLP_LOGS_PROTOCOL    | The transport protocol to use on OTLP log requests. Options include `grpc`, `http/protobuf`, and `http/json`. Default is `http/protobuf`.                    |
+| OTEL_METRICS_EXPORTER    | Metrics exporter to be used. options are `otlp`, `prometheus`, `console` or `none`.                    |
+| OTEL_METRIC_EXPORT_INTERVAL    | The export interval when using a push Metric Reader. Default is `60000`..                    |
+| OTEL_METRIC_EXPORT_TIMEOUT    | The export timeout when using a push Metric Reader. Default is `30000`..                    |
```
```suggestion
| OTEL_METRIC_EXPORT_INTERVAL    | The export interval when using a push Metric Reader. Default is `60000`.                     |
| OTEL_METRIC_EXPORT_TIMEOUT    | The export timeout when using a push Metric Reader. Default is `30000`.                     |
```

bhaskarbanerjee replied:
done

pichlermarc commented on experimental/packages/opentelemetry-sdk-node/src/sdk.ts:
```diff
@@ -86,6 +96,108 @@ export type LoggerProviderConfig = {
+      `OTEL_METRICS_EXPORTER contains "none". Metric provider will not be initialized.`
+    );
+    return metricReaders;
+  } else {
```
we return in the `if` block above so we can reduce nesting here by dropping the else.

bhaskarbanerjee replied:
agree

pichlermarc reviewed (changes requested):
Thanks for addressing the comments - almost there. 

We also need to add the now used exporters as dependencies to `@opentelemetry/sdk-node`'s `package.json`. Please be sure to follow up with a `npm install --package-lock-only` after you do that and ensure that the changes made to `package-lock.json` are added to the commit. 🙂

bhaskarbanerjee commented:
> Thanks for addressing the comments - almost there.
> 
> We also need to add the now used exporters as dependencies to `@opentelemetry/sdk-node`'s `package.json`. Please be sure to follow up with a `npm install --package-lock-only` after you do that and ensure that the changes made to `package-lock.json` are added to the commit. 🙂

pichlermarc reviewed (commented):
Thanks, please also add an entry in `experimental/CHANGELOG.md` then this should be good to merge 🙂

bhaskarbanerjee commented:
@pichlermarc  This is the output of `npm install --package-lock-only` on my system
```
abcabc@mymac opentelemetry-js % nvm use 18                     
Now using node v18.20.2 (npm v10.5.0)
abcabc@mymac opentelemetry-js % npm install --package-lock-only
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'glob@11.0.0',
npm WARN EBADENGINE   required: { node: '20 || >=22' },
npm WARN EBADENGINE   current: { node: 'v18.20.2', npm: '10.5.0' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'minimatch@10.0.1',
npm WARN EBADENGINE   required: { node: '20 || >=22' },
npm WARN EBADENGINE   current: { node: 'v18.20.2', npm: '10.5.0' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'jackspeak@4.0.2',
npm WARN EBADENGINE   required: { node: '20 || >=22' },
npm WARN EBADENGINE   current: { node: 'v18.20.2', npm: '10.5.0' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'path-scurry@2.0.0',
npm WARN EBADENGINE   required: { node: '20 || >=22' },
npm WARN EBADENGINE   current: { node: 'v18.20.2', npm: '10.5.0' }
npm WARN EBADENGINE }
npm WARN EBADENGINE Unsupported engine {
npm WARN EBADENGINE   package: 'lru-cache@11.0.2',
npm WARN EBADENGINE   required: { node: '20 || >=22' },
npm WARN EBADENGINE   current: { node: 'v18.20.2', npm: '10.5.0' }
npm WARN EBADENGINE }
(node:69826) Warning: Setting the NODE_TLS_REJECT_UNAUTHORIZED environment variable to '0' makes TLS connections and HTTPS requests insecure by disabling certificate verification.
(Use `node --trace-warnings ...` to show where the warning was created)

> opentelemetry@0.1.0 postinstall
> npm run update-ts-configs


> opentelemetry@0.1.0 update-ts-configs
> node scripts/update-ts-configs.js


up to date, audited 2355 packages in 6s

224 packages are looking for funding
  run `npm fund` for details

66 vulnerabilities (5 low, 8 moderate, 18 high, 35 critical)

To address issues that do not require attention, run:
  npm audit fix

To address all issues possible (including breaking changes), run:
  npm audit fix --force

Some issues need review, and may require choosing
a different dependency.

Run `npm audit` for details.
abcabc@mymac opentelemetry-js % git diff                       
diff --git a/experimental/packages/opentelemetry-sdk-node/tsconfig.json b/experimental/packages/opentelemetry-sdk-node/tsconfig.json
index 7d18aac48..07232313f 100644
--- a/experimental/packages/opentelemetry-sdk-node/tsconfig.json
+++ b/experimental/packages/opentelemetry-sdk-node/tsconfig.json
@@ -60,18 +60,6 @@
     {
       "path": "../exporter-trace-otlp-proto"
     },
-    {
-      "path": "../opentelemetry-exporter-metrics-otlp-grpc"
-    },
-    {
-      "path": "../opentelemetry-exporter-metrics-otlp-http"
-    },
-    {
-      "path": "../opentelemetry-exporter-metrics-otlp-proto"
-    },
-    {
-      "path": "../opentelemetry-exporter-prometheus"
-    },
     {
       "path": "../opentelemetry-instrumentation"
     },

```
